src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/ShardStateMetricsProcessor.java

```diff
@@ -0,0 +1,94 @@
+    private BatchBindStep handle;
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final TypeReference<HashMap<String, String>> TYPE_REF = new TypeReference<HashMap<String, String>>() {};
+
+    private ShardStateMetricsProcessor(ShardStateMetricsSnapshot snapshot) {
+        this.shardStateMetricsSnapshot = snapshot;
+    }
+
+    static ShardStateMetricsProcessor buildShardStateMetricEventsProcessor(
+            long currWindowStartTime,
+            Connection conn,
+            NavigableMap<Long, ShardStateMetricsSnapshot> shardStateEventMetricsMap) {
+        ShardStateMetricsSnapshot shardStateSnap = shardStateEventMetricsMap.get(currWindowStartTime);
+        if (shardStateSnap == null) {
+            shardStateSnap = new ShardStateMetricsSnapshot(conn, currWindowStartTime);
+            shardStateEventMetricsMap.put(currWindowStartTime, shardStateSnap);
+        }
+        return new ShardStateMetricsProcessor(shardStateSnap);
+    }
+
+    @Override
+    public void initializeProcessing(long startTime, long endTime) {
+        this.handle = shardStateMetricsSnapshot.startBatchPut();
+    }
+
+    @Override
+    public void finalizeProcessing() {
+        if (handle.size() > 0) {
+            handle.execute();
+        }
+        LOG.debug("Final ShardStateEvents metrics {}", shardStateMetricsSnapshot.fetchAll());
+    }
+
+    @Override
+    public void processEvent(Event event) {
+        String[] lines = event.value.split(System.lineSeparator());
+        for (String line : lines) {
+            Map<String, String> shardStateMap = extractEntryData(line);
+            if (!shardStateMap.containsKey(PerformanceAnalyzerMetrics.METRIC_CURRENT_TIME)) {
+                handle.bind(
+                        shardStateMap.get(AllMetrics.ShardStateDimension.INDEX_NAME.toString()),
+                        shardStateMap.get(AllMetrics.ShardStateDimension.SHARD_ID.toString()),
+                        shardStateMap.get(AllMetrics.ShardStateDimension.SHARD_TYPE.toString()),
+                        shardStateMap.get(AllMetrics.ShardStateDimension.NODE_NAME.toString()),
+                        Integer.parseInt(shardStateMap.get(AllMetrics.ShardStateValue.SHARD_STATE_ACTIVE.toString())),
+                        Integer.parseInt(shardStateMap.get(AllMetrics.ShardStateValue.SHARD_STATE_INITIALIZING.toString())),
+                        Integer.parseInt(shardStateMap.get(AllMetrics.ShardStateValue.SHARD_STATE_UNASSIGNED.toString())));
+            }
+        }
+    }
+
+    @Override
+    public boolean shouldProcessEvent(Event event) {
+        return event.key.contains(PerformanceAnalyzerMetrics.sShardStatePath);
+    }
+
+    @Override
+    public void commitBatchIfRequired() {
+        if (handle.size() > BATCH_LIMIT) {
+            handle.execute();
+            handle = shardStateMetricsSnapshot.startBatchPut();
+        }
+    }
+
+    static Map<String, String> extractEntryData(String line) {
```

Comment:
We can add a doc comment stating how a sample line looks like, what values we are expecting and so on ?

Reply:
Done